cli/internal/run/real_run.go

```diff
@@ -89,30 +95,68 @@ func RealRun(
 		Concurrency: rs.Opts.runOpts.Concurrency,
 	}
 
-	mu := sync.Mutex{}
+	logMutex := sync.Mutex{}
+	logWaitGroup := sync.WaitGroup{}
+
+	outputLogsFunc := func(grouped bool, defaultOutWriter io.Writer, defaultErrWriter io.Writer, outBuf bytes.Buffer, errBuf bytes.Buffer) {
+		if grouped {
+			logMutex.Lock()
+			_, outErr := defaultOutWriter.Write(outBuf.Bytes())
+			_, errErr := defaultErrWriter.Write(errBuf.Bytes())
+			logMutex.Unlock()
+			if outErr != nil || errErr != nil {
+				base.UI.Error("Failed to write part of the output to terminal")
+			}
+		}
+		logWaitGroup.Done()
+	}
+
+	taskSummaryMutex := sync.Mutex{}
 	taskSummaries := []*runsummary.TaskSummary{}
 	execFunc := func(ctx gocontext.Context, packageTask *nodes.PackageTask, taskSummary *runsummary.TaskSummary) error {
-		taskExecutionSummary, err := ec.exec(ctx, packageTask)
-		if err != nil {
-			return err
+
+		grouped := rs.Opts.runOpts.LogOrder == "grouped"
+
+		outBuf := &bytes.Buffer{}
+		errBuf := &bytes.Buffer{}
+
+		var outWriter io.Writer
+		var errWriter io.Writer
+
+		defaultOutWriter := os.Stdout
+		defaultErrWriter := os.Stderr
+
+		if grouped {
+			outWriter = outBuf
+			errWriter = errBuf
+		} else {
+			outWriter = defaultOutWriter
+			errWriter = defaultErrWriter
 		}
 
+		ui := concurrentUIFactory.Build(os.Stdin, outWriter, errWriter)
+
+		taskExecutionSummary, err := ec.exec(ctx, packageTask, ui, outWriter, errWriter)
+
 		// taskExecutionSummary will be nil if the task never executed
 		// (i.e. if the workspace didn't implement the script corresponding to the task)
 		// We don't need to collect any of the outputs or execution if the task didn't execute.
-		if taskExecutionSummary != nil {
+		if err == nil && taskExecutionSummary != nil {
 			taskSummary.ExpandedOutputs = taskHashTracker.GetExpandedOutputs(taskSummary.TaskID)
 			taskSummary.Execution = taskExecutionSummary
 			taskSummary.CacheState = taskHashTracker.GetCacheStatus(taskSummary.TaskID)
 
 			// lock since multiple things to be appending to this array at the same time
-			mu.Lock()
+			taskSummaryMutex.Lock()
 			taskSummaries = append(taskSummaries, taskSummary)
 			// not using defer, just release the lock
-			mu.Unlock()
+			taskSummaryMutex.Unlock()
 		}
 
-		return nil
+		logWaitGroup.Add(1)
+		go outputLogsFunc(grouped, defaultOutWriter, defaultErrWriter, *outBuf, *errBuf)
```

Comment:
This is unfortunately still a race condition. There is no guarantee from the Go runtime that _this_ goroutine will start or get the lock before this same goroutine from a downstream task. Fortunately, I think a relatively simple fix will help. What about creating a channel for a struct that bundles all of the necessary data for outputting? The channel has the added bonus of providing ordering.

Then, goroutine that does the actual outputting can loop over the channel using a for loop, which will terminate when the channel is closed.

One thing to watch out for: we'll need to use a buffered channel so that sending on the channel doesn't block. An appropriate buffer size would likely be something like the # of all tasks to be run, just to be safe.

Let me know if this makes sense, or if you would like help with this piece.

Reply:
I am currently unfamiliar with channels, I am sure I will be able to figure it out though. If I can't figure it out I'll let you know.

Reply:
Channels look pretty cool, thanks for making me explore this direction.